python/pyiceberg/expressions/visitors.py

```diff
@@ -842,3 +842,49 @@ def visit_bound_predicate(self, predicate: BoundPredicate[L]) -> Set[int]:
 
 def extract_field_ids(expr: BooleanExpression) -> Set[int]:
     return visit(expr, _ExpressionFieldIDs())
+
+
+class _RewriteToDNF(BooleanExpressionVisitor[BooleanExpression]):
+    def visit_true(self) -> BooleanExpression:
+        return AlwaysTrue()
+
+    def visit_false(self) -> BooleanExpression:
+        return AlwaysFalse()
+
+    def visit_not(self, child_result: BooleanExpression) -> BooleanExpression:
+        return ~child_result
+
+    def visit_and(self, left_result: BooleanExpression, right_result: BooleanExpression) -> BooleanExpression:
+        if isinstance(left_result, Or) and isinstance(right_result, Or):
+            # Distributive law: ((P OR Q) AND (R OR S)) AND (((P AND R) OR (P AND S)) OR ((Q AND R) OR ((Q AND S)))
+            return Or(
+                And(left_result.left, right_result.left),
+                And(left_result.left, right_result.right),
+                And(left_result.right, right_result.left),
+                And(left_result.right, right_result.right),
+            )
+        if isinstance(right_result, Or):
+            # Distributive law: A AND (B OR C) = (A AND B) OR (A AND C)
+            return Or(And(left_result, right_result.left), And(left_result, right_result.right))
+        elif isinstance(left_result, Or):
```

Comment:
What if there are more levels of nesting than this?

Reply:
Since we go from the leaves, this should go well. See:
```python
def test_to_dnf_nested_or() -> None:
    expr = Or(EqualTo("P", "a"), And(EqualTo("Q", "b"), Or(EqualTo("R", "c"), EqualTo("S", "d"))))
    assert rewrite_to_dnf(expr) == (
        EqualTo("P", "a"),
        And(EqualTo("Q", "b"), EqualTo("R", "c")),
        And(EqualTo("Q", "b"), EqualTo("S", "d")),
    )
```